pennylane_lightning/src/simulator/DynamicDispatcher.cpp

```diff
@@ -0,0 +1,184 @@
+// Copyright 2021 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file
+ * Defines
+ */
+#include "DynamicDispatcher.hpp"
+
+#include <unordered_set>
+
+/// @cond DEV
+namespace {
+using Pennylane::DynamicDispatcher;
+using Pennylane::GateOperations;
+using Pennylane::KernelType;
+using Pennylane::SelectGateOps;
+
+using Pennylane::static_lookup;
+
+using Pennylane::AVAILABLE_KERNELS;
+using Pennylane::GATE_NAMES;
+using Pennylane::GATE_NUM_PARAMS;
+} // namespace
+
+#define PENNYLANE_APPLY_OPS_TO_LAMBDA(GATE_NAME)                               \
```

Comment:
Unfortunately, this is the best I got after trying several different approaches. Another method I have tried is making a big switch case statement that returns a correct lambda function for a given GateOperation. See https://github.com/PennyLaneAI/pennylane-lightning/blob/367818d84e7d9d473a54bff221aaf8b1a8902eca/pennylane_lightning/src/simulator/DynamicDispatcher.cpp

Reply:
I guess this is one of those situations that is difficult/impossible to replicate easily, since we are doing source transformations/generation with the preprocessor. Macros are probably the easiest way to accomplish something like this.

I do worry about debugability of this though, as the outputted source will be different than the generating source, which may make it hard/impossible to map debugger lines to the failing routine, if needed. I may need to think about this a bit more too. Maybe it is fine.

Reply:
All preprocessors in this file is replaced with a type of template metaprogramming.